File: api/animation_generator.py
"""
Animation Generator Service
Extracts bone structure from GLB and generates Three.js animation code using OpenRouter
"""

import os
import json
import requests
from typing import Dict, List, Optional, Any
from dotenv import load_dotenv

load_dotenv()

OPENROUTER_API_KEY = os.getenv("OPENROUTER_API_KEY")
OPENROUTER_BASE_URL = "https://openrouter.ai/api/v1"

class AnimationGenerator:
    def __init__(self):
        self.api_key = OPENROUTER_API_KEY
        if not self.api_key:
            raise ValueError("OPENROUTER_API_KEY not set in environment")
        
        self.headers = {
            "Authorization": f"Bearer {self.api_key}",
            "Content-Type": "application/json",
            "HTTP-Referer": "https://osce-webxr-poc.vercel.app",
            "X-Title": "OSCE WebXR Animation Generator"
        }
    
    def extract_bone_structure(self, bone_data: List[Dict]) -> str:
        """
        Format bone structure data for LLM prompt
        bone_data should be a list of dicts with: name, parent, position, rotation
        """
        formatted = json.dumps(bone_data, indent=2)
        return formatted
    
    def generate_animation_code(
        self, 
        bone_structure: str, 
        prompt: str,
        model: str = "anthropic/claude-3.5-sonnet"
    ) -> Dict[str, Any]:
        """
        Generate Three.js animation code from text prompt using OpenRouter
        
        Args:
            bone_structure: JSON string of bone hierarchy
            prompt: Natural language description of desired animation
            model: OpenRouter model identifier
            
        Returns:
            Dict with generated code and metadata
        """
        system_prompt = """You are a Three.js animation expert specializing in skeletal animation.

Given a bone hierarchy from a GLB model, generate executable Three.js code that animates the character based on the user's description.

Requirements:
1. Return ONLY executable JavaScript code, no markdown, no explanations
2. Use skeleton.getBoneByName('BoneName') to access bones
3. Use realistic human joint rotation limits (typically ±45-90 degrees for most joints)
4. For smooth animations, use THREE.AnimationClip or keyframe tracks
5. Return code that can be executed with: new Function('skeleton', 'THREE', code)(skeleton, THREE)
6. If the animation should loop or have duration, include that in the code
7. Use radians for rotations (Math.PI / 2 = 90 degrees)
8. Consider the bone hierarchy - child bones inherit parent transformations

Example output format:
```javascript
// Animation: [description]
const leftUpLeg = skeleton.getBoneByName('LeftUpLeg');
const rightUpLeg = skeleton.getBoneByName('RightUpLeg');
const spine = skeleton.getBoneByName('Spine');

// Create keyframe tracks for smooth animation
const duration = 2.0; // seconds
const times = [0, 1, 2];
const leftLegRotations = [0, Math.PI * 0.3, Math.PI * 0.4];
const rightLegRotations = [0, -Math.PI * 0.3, -Math.PI * 0.4];

// Apply rotations
leftUpLeg.rotation.z = Math.PI * 0.4;
rightUpLeg.rotation.z = -Math.PI * 0.4;
// Lower hips slightly
if (spine) spine.position.y -= 0.3;
```

Generate code that is safe, realistic, and follows Three.js best practices."""

        user_prompt = f"""Bone hierarchy:
{bone_structure}

Generate Three.js code to make this character: "{prompt}"

Return only the executable JavaScript code."""

        payload = {
            "model": model,
            "messages": [
                {"role": "system", "content": system_prompt},
                {"role": "user", "content": user_prompt}
            ],
            "temperature": 0.7,
            "max_tokens": 500  # Reduced to work with limited credits
        }
        
        try:
            response = requests.post(
                f"{OPENROUTER_BASE_URL}/chat/completions",
                headers=self.headers,
                json=payload,
                timeout=30
            )
            response.raise_for_status()
            
            result = response.json()
            generated_code = result["choices"][0]["message"]["content"]
            
            # Clean up code (remove markdown if present)
            code = generated_code.strip()
            if code.startswith("```javascript"):
                code = code.replace("```javascript", "").replace("```js", "").replace("```", "").strip()
            elif code.startswith("```"):
                code = code.replace("```", "").strip()
            
            return {
                "code": code,
                "model": model,
                "prompt": prompt,
                "success": True
            }
            
        except Exception as e:
            return {
                "code": None,
                "error": str(e),
                "success": False
            }
    
    def validate_code(self, code: str) -> Dict[str, Any]:
        """
        Basic validation that code is safe to execute
        (In production, use a sandbox or more sophisticated validation)
        """
        # Check for dangerous patterns
        dangerous_patterns = [
            "eval(",
            "Function(",
            "require(",
            "import(",
            "fetch(",
            "XMLHttpRequest",
            "document.",
            "window.",
            "process."
        ]
        
        for pattern in dangerous_patterns:
            if pattern in code:
                return {
                    "valid": False,
                    "reason": f"Potentially unsafe pattern detected: {pattern}"
                }
        
        # Check that code uses skeleton and THREE
        if "skeleton" not in code and "getBoneByName" not in code:
            return {
                "valid": False,
                "reason": "Code must use skeleton.getBoneByName() to access bones"
            }
        
        return {
            "valid": True,
            "reason": "Code appears safe"
        }

# Global instance
animation_generator = AnimationGenerator() if OPENROUTER_API_KEY else None

File: frontend/src/animation-generator.ts
/**
 * Animation Generator Service
 * Extracts bone structure from GLB models and generates animations via AI
 */

import * as THREE from 'three';
import type { GLTF } from 'three/examples/jsm/loaders/GLTFLoader.js';
import { APIClient } from './api.js';

export interface BoneInfo {
    name: string;
    parent: string | null;
    position: [number, number, number];
    rotation: [number, number, number];
}

export interface AnimationResult {
    code: string;
    validation: {
        valid: boolean;
        reason: string;
    };
    prompt: string;
    model: string;
}

export class AnimationGenerator {
    private apiClient: APIClient;
    
    constructor(apiClient: APIClient) {
        this.apiClient = apiClient;
    }
    
    /**
     * Extract bone structure from a loaded GLTF model
     */
    extractBoneStructure(gltf: GLTF): BoneInfo[] {
        const bones: BoneInfo[] = [];
        
        // First, try to get skeleton from SkinnedMesh
        let skeleton: THREE.Skeleton | null = null;
        gltf.scene.traverse((child) => {
            if (child instanceof THREE.SkinnedMesh && child.skeleton) {
                skeleton = child.skeleton;
            }
        });
        
        // If we found a skeleton, extract bones from it
        if (skeleton !== null) {
            const skeletonBones = (skeleton as any).bones as THREE.Bone[] | undefined;
            if (skeletonBones && Array.isArray(skeletonBones) && skeletonBones.length > 0) {
                skeletonBones.forEach((bone: THREE.Bone) => {
                    bones.push({
                        name: bone.name || 'UnnamedBone',
                        parent: bone.parent?.name || null,
                        position: bone.position.toArray() as [number, number, number],
                        rotation: [
                            bone.rotation.x,
                            bone.rotation.y,
                            bone.rotation.z
                        ]
                    });
                });
            }
        }
        
        // Fallback: if no bones found from skeleton, try direct traversal
        if (bones.length === 0) {
            // Fallback: traverse scene for bones directly
            gltf.scene.traverse((child) => {
                if (child instanceof THREE.Bone || (child as any).isBone) {
                    const bone = child as THREE.Bone;
                    bones.push({
                        name: bone.name || 'UnnamedBone',
                        parent: bone.parent?.name || null,
                        position: bone.position.toArray() as [number, number, number],
                        rotation: [
                            bone.rotation.x,
                            bone.rotation.y,
                            bone.rotation.z
                        ]
                    });
                }
            });
        }
        
        return bones;
    }
    
    /**
     * Get skeleton from a GLTF model
     */
    getSkeleton(gltf: GLTF): THREE.Skeleton | null {
        let skeleton: THREE.Skeleton | null = null;
        
        gltf.scene.traverse((child) => {
            if (child instanceof THREE.SkinnedMesh) {
                skeleton = child.skeleton;
            }
        });
        
        return skeleton;
    }
    
    /**
     * Generate animation code from text prompt
     */
    async generateAnimation(
        boneStructure: BoneInfo[],
        prompt: string,
        model: string = "anthropic/claude-3.5-sonnet"
    ): Promise<AnimationResult> {
        return await this.apiClient.generateAnimation({
            bone_structure: boneStructure,
            prompt: prompt,
            model: model
        });
    }
    
    /**
     * Execute generated animation code safely
     * Creates a sandboxed function that only has access to skeleton and THREE
     */
    executeAnimationCode(
        code: string,
        skeleton: THREE.Skeleton,
        THREE: typeof import('three')
    ): void {
        try {
            // Create a function with limited scope
            const animationFunction = new Function(
                'skeleton',
                'THREE',
                `
                // Helper function to get bone by name
                skeleton.getBoneByName = function(name) {
                    for (let i = 0; i < this.bones.length; i++) {
                        if (this.bones[i].name === name) {
                            return this.bones[i];
                        }
                    }
                    return null;
                };
                
                ${code}
                `
            );
            
            animationFunction(skeleton, THREE);
        } catch (error) {
            console.error('Error executing animation code:', error);
            throw new Error(`Animation execution failed: ${error instanceof Error ? error.message : 'Unknown error'}`);
        }
    }
    
    /**
     * Create an AnimationClip from generated code
     * This is a more advanced approach that creates reusable animations
     */
    createAnimationClip(
        code: string,
        skeleton: THREE.Skeleton,
        duration: number = 2.0,
        name: string = "GeneratedAnimation"
    ): THREE.AnimationClip | null {
        // For now, we'll execute the code directly
        // In the future, we could parse the code to extract keyframe data
        // and create a proper AnimationClip
        
        try {
            this.executeAnimationCode(code, skeleton, THREE);
            
            // Create a simple clip that holds the current pose
            // This is a placeholder - ideally the AI would generate keyframe tracks
            const tracks: THREE.KeyframeTrack[] = [];
            
            // Extract bone rotations from executed code
            skeleton.bones.forEach((bone, index) => {
                if (bone.rotation.x !== 0 || bone.rotation.y !== 0 || bone.rotation.z !== 0) {
                    // Create rotation track
                    const times = [0, duration];
                    const values = [
                        bone.rotation.x, bone.rotation.y, bone.rotation.z,
                        bone.rotation.x, bone.rotation.y, bone.rotation.z
                    ];
                    
                    const track = new THREE.QuaternionKeyframeTrack(
                        `.bones[${index}].quaternion`,
                        times,
                        values
                    );
                    tracks.push(track);
                }
            });
            
            if (tracks.length === 0) {
                return null;
            }
            
            return new THREE.AnimationClip(name, duration, tracks);
        } catch (error) {
            console.error('Error creating animation clip:', error);
            return null;
        }
    }
}
